components/api_request_helper/api_request_helper.cc

```diff
@@ -355,25 +326,54 @@ void APIRequestHelper::URLLoaderHandler::OnDataReceived(
 
 void APIRequestHelper::URLLoaderHandler::OnComplete(bool success) {
   DCHECK(data_completed_callback_);
-  DVLOG(1) << "[[" << __func__ << "]]"
-           << " Response completed\n";
+  VLOG(1) << "[[" << __func__ << "]]"
+          << " Response completed\n";
 
   std::move(data_completed_callback_)
-      .Run(api_request_helper_->ToAPIRequestResult(std::move(url_loader_)),
-           success);
-
-  api_request_helper_->Cancel(this);
+      .Run(ToAPIRequestResult(std::move(url_loader_)), success);
+  // If we're still parsing, we don't want to Cancel as that will delete
+  // the decoder instance for this request. Wait until all decoder operations
+  // are complete.
+  DCHECK_LT(-1, current_decoding_operation_count_);
+  if (current_decoding_operation_count_ == 0) {
+    // Not waiting, mark as done
+    api_request_helper_->Cancel(this);
+    return;
+  }
+  VLOG(1) << "Still have " << current_decoding_operation_count_
+          << " decoding operations in progress, waiting for them to"
+          << " complete...";
+  // Only instantiate this callback during/after OnComplete as
+  // we won't know the final number of chunks until then.
+  data_chunk_parsed_callback_ = base::BindRepeating(
+      [](base::WeakPtr<APIRequestHelper::URLLoaderHandler> instance) {
+        if (instance && instance->current_decoding_operation_count_ == 0) {
+          VLOG(1) << "All decoding operations completed for this request.";
+          instance->api_request_helper_->Cancel(instance.get());
```

Comment:
IMHO better to use base::RefCounted<SomeSubClass> (non thread-safe) to count the active operations automatically.
`Cancel()` could be called automatically in dtor.

Reply:
If it looks like hard to implement I'm okay with the current approach as a fast fix to deal with nightly perf degradation.

Reply:
Frankly speaking, I would like more advanced instance reusing to be implemented. Not only inside one `ApiHelperInstance`, but across all the instances for the same profile. Of course we couldn't do it by default and need to implement a policy for that (for example allow reusing between the same origin only).
Also we called `ParseJsonIsolated` before and don't see this problem.

Therefore I'm not sure that we need `current_decoding_operation_count_` or `base::RefCounted<Something>` right now. If moving ` data_decoder_ = std::make_unique<data_decoder::DataDecoder>()` out of ctor is enough lets do only this for now.

Reply:
It is not enough. Streaming responses need to keep the same reference in order to guarantee response order. This change continues to support that and makes non-streaming responses handled the same as ParseJsonIsolated.

Reply:
> @petemill According to spec `ParseJsonIsolated` always creates a new instance of DataDecoder service.
> IMHO the main problem is creating `Decoder Service` instances in ctor and holding them forever.

Exactly. That's why I moved it from the constructor. Further, I moved it back to being per-request which is inline with the previous code. If we want to change that we can with a security review. I'm also not sure what holding on to the reference for the lifetime of the request helper would do. Presumably it would delete the extra process after inactivity but I didn't research it. Either way, I'd prefer if someone proposes that change in a subsequent PR with a security review.

Reply:
> long periods of time
and
> may be killed

That's too much uncertainty IMO for this PR. It can be proposed in a subsequent one and this one maintains the currently approved methodology and doesn't increase or reduce the number of processes, nor does it change the process creation time or life duration.

Reply:
As discussed over DM, RefCounted would require another layer of complexity in being able to Cancel the reponse when APIRequestHelper is deleted, so we'll keep the counter but avoid the extra callback.